### ISSUE TITLE
BBC: Blead (30fc7a28) Breaks POE

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17175]
Migrated from [rt.perl.org#134482](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134482) (status was 'open')

Searchable as RT134482$

**Comment by p5pRT** [comment]
### From carlos@carlosguevara.com



Please see http&#8203;://matrix\.cpantesters\.org/?dist=POE \.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.31.5:

Configured by cpan at Fri Oct  4 21:40:08 CDT 2019.

Summary of my perl5 (revision 5 version 31 subversion 5) configuration:
  Commit id: 393dcb6c04b49721349bf7dc3ca104cf64600045
  Platform:
    osname=freebsd
    osvers=11.2-release-p14
    archname=amd64-freebsd
    uname='freebsd cjg-freebsd11 11.2-release-p14 freebsd
11.2-release-p14 #0: mon aug 19 22:38:50 utc 2019
root@amd64-builder.daemonology.net:usrobjusrsrcsysgeneric amd64 '
    config_args='-des -Dprefix=~/bin/perl-blead
-Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H
-fno-strict-aliasing -pipe -fstack-protector-strong
-I/usr/local/include -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H
-fno-strict-aliasing -pipe -fstack-protector-strong
-I/usr/local/include'
    ccversion=''
    gccversion='4.2.1 Compatible FreeBSD Clang 6.0.0
(tags/RELEASE_600/final 326565)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-Wl,-E  -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib /usr/lib/clang/6.0.0/lib /usr/lib
    libs=-lpthread -lgdbm -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags='-DPIC -fPIC'
    lddlflags='-shared -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.31.5:
    /home/cpan/bin/perl-blead/lib/site_perl/5.31.5/amd64-freebsd
    /home/cpan/bin/perl-blead/lib/site_perl/5.31.5
    /home/cpan/bin/perl-blead/lib/5.31.5/amd64-freebsd
    /home/cpan/bin/perl-blead/lib/5.31.5


Environment for perl 5.31.5:
    HOME=/home/cpan
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl-blead/bin:/home/cpan/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/home/cpan/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat\, 05 Oct 2019 16&#8203;:40&#8203;:05 GMT\, carlos@&#8203;carlosguevara\.com wrote&#8203;:
> This is a bug report for perl from "Carlos Guevara"
> \<carlos@&#8203;carlosguevara\.com>\,
> generated with the help of perlbug 1\.41 running under perl 5\.31\.5\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> Please see http&#8203;://matrix\.cpantesters\.org/?dist=POE \.
> 

The test failure looks like this&#8203;:

\#\#\#\#\#
\[POE\-1\.367\] 554 $ bleadprove \-vb t/90\_regression/leolo\-sig\-die\.t
t/90\_regression/leolo\-sig\-die\.t \.\. 
1\.\.11
Constants from lexical variables potentially modified elsewhere are no longer permitted at /home/jkeenan/\.cpanm/work/1570295901\.21531/POE\-1\.367/blib/lib/POE/Kernel\.pm line 368\.
BEGIN failed\-\-compilation aborted at /home/jkeenan/\.cpanm/work/1570295901\.21531/POE\-1\.367/blib/lib/POE/Kernel\.pm line 397\.
Compilation failed in require at \(eval 10\) line 1\.
BEGIN failed\-\-compilation aborted at \(eval 10\) line 1\.
could not import qw\(Kernel\) at t/90\_regression/leolo\-sig\-die\.t line 11\.
BEGIN failed\-\-compilation aborted at t/90\_regression/leolo\-sig\-die\.t line 11\.
\# Looks like your test exited with 255 before it could output anything\.
Dubious\, test returned 255 \(wstat 65280\, 0xff00\)
Failed 11/11 subtests 

Test Summary Report


---
t/90\_regression/leolo\-sig\-die\.t \(Wstat&#8203;: 65280 Tests&#8203;: 0 Failed&#8203;: 0\)
&nbsp;  Non\-zero exit status&#8203;: 255
&nbsp;  Parse errors&#8203;: Bad plan\.  You planned 11 tests but ran 0\.
Files=1\, Tests=0\,  1 wallclock secs \( 0\.04 usr  0\.02 sys \+  0\.10 cusr  0\.02 csys =  0\.18 CPU\)
Result&#8203;: FAIL
\#\#\#\#\#

Bisection points to \.\.\.

\#\#\#\#\#
commit 30fc7a2809e5a175e2d9bb94d765b2039f270d91
Author&#8203;:     James E Keenan \<jkeenan@&#8203;cpan\.org>
AuthorDate&#8203;: Sat May 25 21&#8203;:40&#8203;:00 2019 \-0400
Commit&#8203;:     Tony Cook \<tony@&#8203;develop\-help\.com>
CommitDate&#8203;: Wed Oct 2 14&#8203;:59&#8203;:31 2019 \+1000
Eliminate modifiable variables in constants
&nbsp;    
Transform previously deprecated cases into exceptions\.
&nbsp;    
Update diagnostic; change D to F
&nbsp;    
remove now irrelevant code \(TonyC\)
&nbsp;    
For&#8203;: RT 134138
\#\#\#\#\#

The commit in question implemented the scheduled fatalization in perl\-5\.32 of a situation which was previously deprecated\.  So\, other things being equal\, the CPAN code should be updated to adapt to the changes in core\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat\, 05 Oct 2019 17&#8203;:24&#8203;:49 GMT\, jkeenan wrote&#8203;:
> On Sat\, 05 Oct 2019 16&#8203;:40&#8203;:05 GMT\, carlos@&#8203;carlosguevara\.com wrote&#8203;:
> > This is a bug report for perl from "Carlos Guevara"
> > \<carlos@&#8203;carlosguevara\.com>\,
> > generated with the help of perlbug 1\.41 running under perl 5\.31\.5\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > Please see http&#8203;://matrix\.cpantesters\.org/?dist=POE \.
> >
> 
> The test failure looks like this&#8203;:
> 
> \#\#\#\#\#
> \[POE\-1\.367\] 554 $ bleadprove \-vb t/90\_regression/leolo\-sig\-die\.t
>  t/90\_regression/leolo\-sig\-die\.t \.\.
> 1\.\.11
> Constants from lexical variables potentially modified elsewhere are no
> longer permitted at /home/jkeenan/\.cpanm/work/1570295901\.21531/POE\-
> 1\.367/blib/lib/POE/Kernel\.pm line 368\.
> BEGIN failed\-\-compilation aborted at
> /home/jkeenan/\.cpanm/work/1570295901\.21531/POE\-
> 1\.367/blib/lib/POE/Kernel\.pm line 397\.
> Compilation failed in require at \(eval 10\) line 1\.
> BEGIN failed\-\-compilation aborted at \(eval 10\) line 1\.
> could not import qw\(Kernel\) at t/90\_regression/leolo\-sig\-die\.t line
> 11\.
> BEGIN failed\-\-compilation aborted at t/90\_regression/leolo\-sig\-die\.t
> line 11\.
> \# Looks like your test exited with 255 before it could output
> anything\.
> Dubious\, test returned 255 \(wstat 65280\, 0xff00\)
>  Failed 11/11 subtests
> 
> Test Summary Report
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> t/90\_regression/leolo\-sig\-die\.t \(Wstat&#8203;: 65280 Tests&#8203;: 0 Failed&#8203;: 0\)
>   Non\-zero exit status&#8203;: 255
>   Parse errors&#8203;: Bad plan\.  You planned 11 tests but ran 0\.
> Files=1\, Tests=0\,  1 wallclock secs \( 0\.04 usr  0\.02 sys \+  0\.10 cusr
> 0\.02 csys =  0\.18 CPU\)
> Result&#8203;: FAIL
> \#\#\#\#\#
> 
> Bisection points to \.\.\.
> 
> \#\#\#\#\#
> commit 30fc7a2809e5a175e2d9bb94d765b2039f270d91
> Author&#8203;:     James E Keenan \<jkeenan@&#8203;cpan\.org>
> AuthorDate&#8203;: Sat May 25 21&#8203;:40&#8203;:00 2019 \-0400
> Commit&#8203;:     Tony Cook \<tony@&#8203;develop\-help\.com>
> CommitDate&#8203;: Wed Oct 2 14&#8203;:59&#8203;:31 2019 \+1000
> Eliminate modifiable variables in constants
> 
> Transform previously deprecated cases into exceptions\.
> 
> Update diagnostic; change D to F
> 
> remove now irrelevant code \(TonyC\)
> 
> For&#8203;: RT 134138
> \#\#\#\#\#
> 
> The commit in question implemented the scheduled fatalization in perl\-
> 5\.32 of a situation which was previously deprecated\.  So\, other things
> being equal\, the CPAN code should be updated to adapt to the changes
> in core\.
> 
> Thank you very much\.

Relevant code from POE&#8203;::Kernel is attached\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

 338 BEGIN \{
 339   \# Shorthand for defining an assert constant\.
 340   sub \_define\_assert \{
 341     no strict 'refs';
 342     foreach my $name \(@&#8203;\_\) \{
 343       next if defined \*\{"ASSERT\_$name"\}\{CODE\};
 344       my $assert\_value = &ASSERT\_DEFAULT;
 345       my $assert\_name  = "ASSERT\_$name";
 346       \*$assert\_name = sub \(\) \{ $assert\_value \};
 347     \}
 348   \}
 349 
 350   \# Assimilate POE\_TRACE\_\* and POE\_ASSERT\_\* environment variables\.
 351   \# Environment variables override everything else\.
 352   while \(my \($var\, $val\) = each %ENV\) \{
 353     next unless $var =~ /^POE\_\(\[A\-Z\_\]\+\)$/;
 354 
 355     my $const = $1;
 356 
 357     next unless $const =~ /^\(?&#8203;:TRACE|ASSERT\)\_/ or do \{ no strict 'refs'; defined &$const \};
 358 
 359     \# Copy so we don't hurt our environment\.
 360     my $value = $val;
 361     \($value\) = \($value =~ /^\(\[\-\\@&#8203;\\w\.\]\+\)$/\); \# Untaint per rt\.cpan\.org 81550
 362     $value =~ tr\['"\]\[\]d;
 363     $value = 0 \+ $value if $value =~ /^\\s\*\-?\\d\+\(?&#8203;:\\\.\\d\+\)?\\s\*$/;
 364 
 365     no strict 'refs';
 366     local $^W = 0;
 367     local $SIG\{\_\_WARN\_\_\} = sub \{ \}; \# redefine
 368     \*$const = sub \(\) \{ $value \};  \#\#\#\#\# \<\- POINT OF FAILURE
 369   \}
 370 
 371   \# TRACE\_FILENAME is special\.
 372   \{
 373     no strict 'refs';
 374     my $trace\_filename = TRACE\_FILENAME\(\) if defined &TRACE\_FILENAME;
 375     if \(defined $trace\_filename\) \{
 376       open $trace\_file\_handle\, ">$trace\_filename"
 377         or die "can't open trace file \`$trace\_filename'&#8203;: $\!";
 378       CORE&#8203;::select\(\(CORE&#8203;::select\($trace\_file\_handle\)\, $| = 1\)\[0\]\);
 379     \}
 380   \}
 381   \# TRACE\_DEFAULT changes the default value for other TRACE\_\*
 382   \# constants\.  Since define\_trace\(\) uses TRACE\_DEFAULT internally\, it
 383   \# can't be used to define TRACE\_DEFAULT itself\.
 384 
 385   defined &TRACE\_DEFAULT or \*TRACE\_DEFAULT = sub \(\) \{ 0 \};
 386 
 387   \_define\_trace qw\(
 388     EVENTS FILES PROFILE REFCNT RETVALS SESSIONS SIGNALS STATISTICS
 389   \);
 390 
 391   \# See the notes for TRACE\_DEFAULT\, except read ASSERT and assert
 392   \# where you see TRACE and trace\.
 393 
 394   defined &ASSERT\_DEFAULT or \*ASSERT\_DEFAULT = sub \(\) \{ 0 \};
 395 
 396   \_define\_assert qw\(DATA EVENTS FILES RETVALS USAGE\);
 397 \}

**Comment by tonycoz** [comment]
Fixed upstream in https://github.com/rcaputo/poe/commit/cf5b772fa54e0cda0e447c33ce351f1e6c074d42 but not released yet.

**Comment by jkeenan** [comment]
> Fixed upstream in [rcaputo/poe@cf5b772](https://github.com/rcaputo/poe/commit/cf5b772fa54e0cda0e447c33ce351f1e6c074d42) but not released yet.

Can anyone contact CPAN maintainer Rocco Caputo or original author Curtis Poe IRL to get a new CPAN release?

Thank you very much.
Jim Keenan

**Comment by tonycoz** [comment]
POE 1.368 has been released with a fix.
